src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy

```diff
@@ -124,13 +127,13 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
     }
 
     void saveImageId() {
-        TaskStateHelper state = new TaskStateHelper(name, project.buildDir)
+        TaskStateHelper state = new TaskStateHelper(this.class.simpleName, project.buildDir)
         state.put("imageId", getImageId())
     }
 
     boolean previouslyBuiltImageExists() {
         boolean imageExists = false
-        String prevImageId = new TaskStateHelper(name, project.buildDir).get("imageId")
+        String prevImageId = new TaskStateHelper(this.class.simpleName, project.buildDir).get("imageId")
```

Comment:
@cdancy Got the limitations by adding test cases. `upToDateWhen` only works if you have single task that stores the task execution history. However, it wouldn't work if you have two tasks of type `DockerBuildImage`. In those cases `upToDate` would still check the history of the task. That's why `onlyIf` is the better choice here.